R/datasets.R

```diff
@@ -0,0 +1,108 @@
+# Copyright 2022 Bedford Freeman & Worth Pub Grp LLC DBA Macmillan Learning.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#' BERT Dataset
+#'
+#' Prepare a dataset for BERT-like models.
+#'
+#' @param x A data.frame with one or more character predictor columns.
+#' @param y A factor of outcomes, or a data.frame with a single factor column.
+#'   Can be NULL (default).
+#' @inheritParams tokenize_bert
+#'
+#' @return An initialized \code{\link[torch]{dataset}}.
+#'
+#' @export
+dataset_bert <- torch::dataset(
+  name = "bert_dataset",
+  initialize = function(x, y = NULL, n_tokens = 128L) {
+    # Eventually this should be exported somewhere. It's a super quick version of
+    # something I'm also implementing in tidybert.
+    stopifnot(all(purrr::map_lgl(x, is.character)))
+
+    tokenized_text <- do.call(
+      tokenize_bert,
+      c(
+        x,
+        list(n_tokens = n_tokens)
+      )
+    )
+
+    y <- .standardize_bert_dataset_outcome(y)
+
+    # We supply the data as tensors.
+    self$tokenized_text <- torch::torch_tensor(tokenized_text$token_ids)
+    self$token_types <- torch::torch_tensor(tokenized_text$token_type_ids)
+
+    # Also supply the labels as tensors.
+    self$y = torch::torch_tensor(as.integer(y))
```

Comment:
Just a note while I'm thinking of it that the labels are often plagued by the 0/1-indexing issue, and eventually we should try to catch possible errors of this type.

Reply:
I hope so, but if I recall there is possibly some inconsistency in torch (at least in the documentation) about this.

From the documentation for `torch::nn_cross_entropy_loss()`:

> This criterion expects a class index in the range [0, C-1] as the target for each value of a 1D tensor of size minibatch; if ignore_index is specified, this criterion also accepts this class index (this index may not necessarily be in the class range).

I _think_ that the actual behavior is 1-indexed in this case, but I haven't checked all the loss functions.

Reply:
Ok, let's definitely watch for this as we use things! I think if something expects 0-indexed labels in {torch}, that's probably a {torch} bug, given the convention of the rest of the package.